### PR TITLE
Build releases in GitHub Actions

### PR DESCRIPTION
**Pull request by thepwagner**
Introduce a workflow file for performing the release process on GitHub Actions. Using a [build service](https://slsa.dev/spec/v0.1/requirements#build-service) is a good idea and a requirement for SLSA level 2.

The trade-off of this change is that release notes need to be pre-assembled: since there isn't an opportunity to "fixup" the entries generated from the commit log. WDYT? Alternatively we could use the generated changelog and "edit release" to fix up, but I dig the 2-person review process that a manual changelog bakes in.

To exercise the new process, this PR also proposes a release. It is in the `"Commit and [send in your changes](#send-in-your-changes), including the reason you selected the target version."` state, from the updated `CONTRIBUTING.md`:
* I'm proposing that https://github.com/grafeas/voucher/compare/v2.6.2...main be released as `v2.7.0`. I chose `v2.7.x` as new functionality is backwards compatible.
* If approved, I'll tag and push `v2.7.0` to `Shopify/voucher` for testing 🤞 
   * If necessary, I'll release additional patch increments in `Shopify/voucher` for testing: `v2.7.1`, `v2.7.2`
   * 🎩 someone else will do a mock release against `Shopify/voucher`
* ⚠️ This will be cherry-picked on to `grafeas/voucher`, and the `v2.7.0` process will be repeated (producing the actual release).

# Related
- https://github.com/grafeas/voucher/issues/41

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -67,22 +67,24 @@ $ git push -u <your username> <the name of your branch>
 are adding code which would be untested, please consider adding tests to cover
 that code.
 
+- For new features or changes users should be aware of, add an entry to `CHANGELOG.md`.
```
Comment (dani-santos-code) [inline CONTRIBUTING.md:70]: should we also add a reminder in a PR template?

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      - name: Extract release changelog
+        run: |
+          version=${GITHUB_REF#refs/tags/v*}
+          mkdir -p tmp
+          sed sed '/^# '$version'/,/^# /!d;//d;/^\s*$/d' CHANGELOG.md > tmp/release_changelog.md
+      - name: Release
+        uses: goreleaser/goreleaser-action@5df302e5e9e4c66310a6b6493a8865b12c555af2
+        with:
+          distribution: goreleaser
+          version: v1.1.0
```
Comment (dani-santos-code) [inline .github/workflows/release.yml]: why `v1.1.0`? I see the [latest version is v1.2.1](https://github.com/goreleaser/goreleaser/)

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+          args: release --rm-dist --release-notes=tmp/release_changelog.md
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Comment (dani-santos-code) [inline .github/workflows/release.yml:35]: how are we going to be able to set secrets on the grafeas org without being admins?

**Review by dani-santos-code — APPROVED** [review]
This is a great idea! We should do the same for kubeaudit!

> I'm proposing that grafeas/voucher@v2.6.2...main be released as v2.7.0. I chose v2.7.x as new functionality is backwards compatible.

Agreed!

> If approved, I'll tag and push v2.7.0 to Shopify/voucher for testing 🤞

🤞

> 🎩 someone else will do a mock release against Shopify/voucher
 
 I can do a mock release

File: .github/workflows/release.yml (hunk 2 above)
Comment (thepwagner) [reply .github/workflows/release.yml]:
```suggestion
          version: v1.2.1
```

I just missed them! `1.2.0` and `1.2.1` were released 15 hours ago, this PR is 20 hours old. 😆 

Let's upgrade, the SBOM stuff in `v1.2.0` looks interesting!

File: .github/workflows/release.yml (hunk 3 above)
Comment (thepwagner) [reply .github/workflows/release.yml:35]: We don't need to, as `GITHUB_TOKEN` is an automatic token scoped to the repository - https://docs.github.com/en/actions/security-guides/automatic-token-authentication

The permissions for this token are what we're setting in the `permissions:` block of the workflow file. The automatic token is only valid while the workflow is running.

File: CONTRIBUTING.md (hunk 1 above)
Comment (dani-santos-code) [reply CONTRIBUTING.md:70]: realized that we currently have no PR templates...if we did, it'd be nice to add a reminder, but..

File: .github/workflows/release.yml (hunk 3 above)
Comment (dani-santos-code) [reply .github/workflows/release.yml:35]: oh, interesting! thanks for the explanation!!

**Comment by thepwagner** [comment]
This is ready to 🎩 !

I made three test releases: https://github.com/Shopify/voucher/releases
* v2.7.0 caught a bug in the `release.yaml` workflow.
* https://github.com/Shopify/voucher/releases/tag/v2.7.1 uses a generated `CHANGELOG.md`, in case we decide against that process.
* https://github.com/Shopify/voucher/releases/tag/v2.7.2 sources release notes from `CHANGELOG.md`

Both releases should have the same artifacts, with roughly the same contents as previous `grafeas/voucher` releases:
```
$ curl -sL https://github.com/grafeas/voucher/releases/download/v2.6.2/voucher_client_2.6.2_Darwin_arm64.tar.gz | tar xvvz
x -rw-r--r--  0 pwagner staff   11357 25 Aug 15:41 voucher_client_2.6.2_Darwin_arm64/LICENSE
x -rwxr-xr-x  0 pwagner staff 8572562 21 Oct 10:11 voucher_client_2.6.2_Darwin_arm64/voucher_client
$ voucher_client_2.6.2_Darwin_arm64/voucher_client --version
voucher_client version 2.6.2

$ curl -sL https://github.com/Shopify/voucher/releases/download/v2.7.2/voucher_client_2.7.2_Darwin_arm64.tar.gz | tar xvvz
x -rw-r--r--  0 runner docker  11357 22 Dec 13:36 voucher_client_2.7.2_Darwin_arm64/LICENSE
x -rwxr-xr-x  0 runner docker 13327570 22 Dec 13:41 voucher_client_2.7.2_Darwin_arm64/voucher_client
$ voucher_client_2.7.2_Darwin_arm64/voucher_client --version
voucher_client version 2.7.2
```

But artifacts are compiled in an ephemeral worker [managed by GitHub Actions](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md#cached-tools), not on the releaser's rando laptop 🎉 .
```
$ strings voucher_client_2.6.2_Darwin_arm64/voucher_client | grep -i pwagner | wc -l
     180
$ strings voucher_client_2.7.2_Darwin_arm64/voucher_client | grep -i pwagner | wc -l
       0
```

**Comment by thepwagner** [comment]
I'm super happy with this solution.
I think the process change of maintaining a `CHANGELOG.md` and reviewing releases is worth buy-in/consensus.

That means parking until January ☃️ !

**Review by cailyn-codes — APPROVED** [review]

**Review by shane-lawrence — APPROVED** [review]

**Review by anthonyho007 — APPROVED** [review]

**Comment by thepwagner** [comment]
Thanks for the feedback on the process change! 🚀 

> This will be cherry-picked on to grafeas/voucher, and the v2.7.0 process will be repeated (producing the actual release).

This became https://github.com/grafeas/voucher/pull/58 upstream

**Pull request closed without merging**